[mlir] Enable decoupling two kinds of greedy behavior.

jpienaar opened the pull request:
The greedy rewriter is used in many different flows and it has a lot of convenience (work list management, debugging actions, tracing, etc). But it combines two kinds of greedy behavior 1) how ops are matched, 2) folding wherever it can.

These are independent forms of greedy and leads to inefficiency. E.g., cases where one need to create different phases in lowering and is required to applying patterns in specific order split across different passes. Using the driver one ends up needlessly retrying folding/having multiple rounds of folding attempts, where one final run would have sufficed.

Of course folks can locally avoid this behavior by just building their own, but this is also a common requested feature that folks keep on working around locally in suboptimal ways.

For downstream users, there should be no behavioral change. Updating from the deprecated should just be a find and replace (e.g., `find ./ -type f -exec sed -i 's|applyPatternsAndFoldGreedily|applyPatternsGreedily|g' {} \;` variety) as the API hasn't changed between the two.

jpienaar committed:
[mlir] Enable decoupling two kinds of greedy behavior.

The greedy rewriter is used in many different flows and it has a lot of
convenience (work list management, debugging actions, tracing, etc). But
it combines two kinds of greedy behavior 1) wrt how ops are matched, 2)
folding wherever it can.

These are independent forms of greedy and leads to inefficiency. E.g.,
cases where one need to create different phases in lowering, one is
required to applying patterns in specific order/different passes. But if
using the driver one ends up needlessly retrying folding or having
multiple rounds of folding attempts, where one final run would have
sufficed. It also is rather confusing to users that just want to apply
some patterns while having all the convenience and structure to have
unrelated changes to IR.

Of course folks can locally avoid this behavior by just building their
own, but this is also a common requested feature that folks keep on
working around locally in suboptimal ways.

joker-eph commented:
> cases where one need to create different phases in lowering and is required to applying patterns in specific order split across different passes

I don’t understand what this means, or how it relates to greedy, or what is the actual problem to solve right now…

> this is also a common requested feature that folks keep on working around locally in suboptimal ways.

I am slightly concerned that this request comes from some wrong design or incorrect root causing.
(The request isn’t common enough that I faced it by the way)

jpienaar commented:
> I don’t understand what this means, or how it relates to greedy, or what is the actual problem to solve right now…

If one has a set of patterns that have to all be applied before another set of patterns, then one needs to split it into multiple passes today (the priorities are not sufficient depending on iteration order). Now one ends up with trying to do the same folding multiple times needlessly vs if you just did it in one. This enables using different sets of patterns relatively cheaply, rather than all checks for folding being wasted until last/if ever. A lot of interface and dialect fallback hook querying until then, all wasted: 38% of total execution time spent in mlir::detail::constant_op_binder::match, of that significant time in hasTrait called from there in the profile I'm looking at. I don't think there is any folding even happening in the end, just a lot of attempts made (e.g., its not expensive folders being an issue, just actually querying values to see if they are constant and a folder can be used is).

jpienaar commented:
To make it more concrete:

All passes/pipelines combined 140s
applyPatternsAndFoldGreedily 105s
- fold 93s
  - constant matching 63s (there are some additional constant matching outside folding, < 0.5s)
     - hasTrait related to 57s

So ~50% of time is spent just checking if an operation is a constant (don't have all the line info to see time spent in repopulating constant hash map etc). The only ways to reduce overhead at the moment is to combine into larger patternsets (only sometimes possible) or not using the rewrite driver (which means you either lose the worklist management & nicer debug/tracing facilities or just create duplication).

matthias-springer commented on mlir/include/mlir/Transforms/GreedyPatternRewriteDriver.h:
```diff
@@ -91,6 +91,15 @@ class GreedyRewriteConfig {
   /// An optional listener that should be notified about IR modifications.
   RewriterBase::Listener *listener = nullptr;
+
+  // Whether this should fold while greedily rewriting.
```
nit: tripple `///`

matthias-springer commented on mlir/lib/Transforms/Utils/GreedyPatternRewriteDriver.cpp:
```diff
@@ -840,6 +840,11 @@ LogicalResult RegionPatternRewriteDriver::simplify(bool *changed) && {
     auto insertKnownConstant = [&](Operation *op) {
+      // This hoisting is to enable more folding, so skip checking if known
+      // constant, updating dense map etc if not doing folding.
+      if (!config.fold)
```
`insertKnownConstant` is not really doing "folding": it does not call the folder. What it does is CSE'ing constants. So I would go as far as adding a second config flag: `cseConstants` (see #89552). Alternatively, update the `GreedyRewriteConfig::fold` documentation and mention that setting it to "false" also disables the CSE'ing of constants.

joker-eph replied:
I am quite concern with the accumulation of flags: this seems to me like working around problems instead of addressing them, and adding complexity that I don't see manageable on the long term.

jpienaar replied:
What problem do you think is being worked around?

For me I see this as I'm a user that wants a worklist driven driver, I have a curated pipeline where I know there isn't any folding here (e.g., the ops being created have no folding, its part of pipeline where folding is ~0% likely, so the ROI on checking if it can be folded is none), but they have no way of communicating that intent. The problem is expert users not being able to express intent without crafting own/local forking.

Now, constant CSE'ing as Matthias correctly points out is different. It is documented here as enabling more folding, so disabled too (this is cheaper here too as all constant hoisting would have happened earlier and there may be no additional chance here to produce additional constants that would need CSE'ing). But it is different so I could see controlling separately is good. And don't mind doing that. I think this is a framework for high performance, so reasonably defaults that can be tuned seems reasonable to me. I considered the maintenance cost but specialized drivers that share most of the same code seemed higher maintenance cost to me.

matthias-springer reviewed (approved):
I think it's a good idea to make the driver more customizable, esp. if it can help with performance. By specifying patterns, users have control over which ops will be processed. But it is currently not possible to control which ops will be visited for folding purposes, so all ops are visited.

River707 commented:
> To make it more concrete:
> 
> All passes/pipelines combined 140s applyPatternsAndFoldGreedily 105s
> 
> * fold 93s
>   
>   * constant matching 63s (there are some additional constant matching outside folding, < 0.5s)
>     
>     * hasTrait related to 57s
> 
> So ~50% of time is spent just checking if an operation is a constant (don't have all the line info to see time spent in repopulating constant hash map etc). The only ways to reduce overhead at the moment is to combine into larger patternsets (only sometimes possible) or not using the rewrite driver (which means you either lose the worklist management & nicer debug/tracing facilities or just create duplication).

Is the hasTrait purely the ConstantLike trait check? or cumulative checking for the various operands/etc.? Feels like there is a lot of low hanging fruit here for optimization, even outside of decoupling folding.

joker-eph reviewed (changes requested):
There aren't any tests.

joker-eph commented on mlir/include/mlir/Transforms/GreedyPatternRewriteDriver.h:
```diff
@@ -91,6 +91,15 @@ class GreedyRewriteConfig {
+  // patterns based purely on benefit and applying without backtracking using
+  // default cost model, 2) greedily folding where possible while attempting to
+  // match and rewrite using the provided patterns. With this option set to
+  // false it only does the former.
```
I don't understand the note?

jpienaar replied:
Which part?

joker-eph commented on mlir/include/mlir/Transforms/GreedyPatternRewriteDriver.h:
```diff
@@ -117,10 +126,18 @@ class GreedyRewriteConfig {
+  config.fold = true;
+  return applyPatternsGreedily(region, patterns, config, changed);
+}
 
```
The API split makes sense, but then the GreedyRewriteConfig changes don't!

jpienaar replied:
The top one means it may fold, this one it must.

I could hide it further down and thread folding flag through along with config if that is preferred.

joker-eph replied:
> The top one means it may fold, this one it must.

The top one with the right config is equivalent to this one, why would we have a separate API here? Why specifically for fold and not for other parameters of the config?

jpienaar replied:
Mostly as it is already widely used and so makes possible to integrate without much fanfare. I can mark it as deprecated for removal post. But it seemed good way to avoid too much effort on downstream for relatively little code.

joker-eph replied:
Yes please mark as deprecated and plan for removal soon.

jpienaar replied:
Done. Should I migrate core users to the non-deprecated form too?

joker-eph replied:
I think we need to: otherwise we'll get warnings in tree.

jpienaar commented:
> Is the hasTrait purely the ConstantLike trait check? or cumulative checking for the various operands/etc.? Feels like there is a lot of low hanging fruit here for optimization, even outside of decoupling folding.

Its cumulative. This is across all pass executions for the pipeline, so there are multiple greedy pattern applications along with other passes/walks. There 3 iterations at most for any of them (2 being minimum if mutation is made). Now, this is the entire pipeline so at the start of pipeline there may be some cleanup value. But none of the dialects matched/produced implement folders, so in this pipeline beyond the first pass, making checking for constants cheaper just makes the wasted time lower (it doesn't really have a benefit, disabling it improves compile time significantly with no negative performance impact).

jpienaar committed:
Address review comments.

This partially incorporates #89552. I haven't exposed it on
canonicalizer pass as that could be distinct discussion.

jpienaar committed:
Add tests & fix call site

jpienaar commented:
> There aren't any tests.

Fixed.

joker-eph commented on mlir/test/Transforms/test-operation-folder.mlir:
```diff
@@ -34,3 +37,37 @@ func.func @test_dont_reorder_constants() -> (i32, i32, i32) {
+  // NOCSE-DAG: arith.constant 2 : i32
+  // NOCSE-DAG: arith.constant 1 : i32
+  // NOCSE-DAG: arith.constant 2 : i32
+  // NOCSE-DAG: arith.constant 3 : i32
```
Is this really the behavior we want here? Seems like we still have some constant folding (more constants are introduced) but they aren't deduped!

jpienaar replied:
I was confused for second too :) This is the no CSE case. So we do have folding, just not constant depupe. This is flag Matthias suggested, so these two features are independent.

joker-eph replied:
The comment for fold says:

```
 /// Whether this should fold while greedily rewriting. This also disables
   /// CSE'ing constants.
   bool fold = true;
```

Why aren't we keeping these independent then? Disabling Fold shouldn't disable CSE if the CSE option is true...

jpienaar replied:
Out of date comment, sorry. It is/should be independent. I'll double check now.

jpienaar committed:
Add deprecation message

jpienaar committed:
Update callsites & comment

kuhar reviewed (commented):
Should we separate the new API (`applyPatternsGreedily`), switching the in-tree pases to use it, and deprecation into it into a few PRs?

I'm worried that landing all of these in one step can results in reverts that we can avoid with a gradual rollout... Especially with lots of folks disappearing for vacation and potentially not being around to investigate breakages in their areas of interest.

MaheshRavishankar commented:
> Should we separate the new API (`applyPatternsGreedily`), switching the in-tree pases to use it, and deprecation into it into a few PRs?
> 
> I'm worried that landing all of these in one step can results in reverts that we can avoid with a gradual rollout... Especially with lots of folks disappearing for vacation and potentially not being around to investigate breakages in their areas of interest.

Won't the fold being set to true by default make this an NFC change?

kuhar commented:
> Won't the fold being set to true by default make this an NFC change?

What tripped me is that in `applyPatternsAndFoldGreedily` this PR explicitly sets it to `true`:
```c++
  config.fold = true;
  return applyPatternsGreedily(region, patterns, config, changed);
}
```

@jpienaar could you add a tl;dr section to the PR description that lists the API-level changes and suggests how to integrate this in downstream projects? The motivation is very clear, but what to do with the PR not necesserily.

aartbik reviewed (approved):
Okay for sparse (assuming fold remains the default, since all tests pass... ;-)

joker-eph commented:
> What tripped me is that in applyPatternsAndFoldGreedily this PR explicitly sets it to true, making me think the default must be false:

No this does not change the default: it is because this API has "Fold" in the name that we override the flag to ensure preserving the existing behavior of this API. Having this override or not is NFC in the current condition though, it seems like mostly "defensive programming" here.

Folks having the deprecated message should not have anything else to do than a mass replace of `applyPatternsAndFoldGreedily` to `applyPatternsGreedily` (all the adjustments upstream in this diff are demonstrating this I believe).

jpienaar commented:
Yes this should be a NFC change. And yes just defense (the default is to fold, but the current API with fold in the name does take an option and so someone could have disabled folding in the config, called function that says it folds and then be surprised - I mean they really shouldn't :)). Yes I updated by simply doing find and replace textually. With current default nothing else needed. Let me update description to capture this.

jpienaar committed:
Update flang side too

jpienaar committed:
clang format

kazutakahirata commented:
With this patch, I am getting:

```
mlir/lib/Conversion/GPUCommon/GPUToLLVMConversion.cpp:535:13: error: 'applyPatternsAndF
oldGreedily' is deprecated: Use applyPatternsGreedily() instead [-Werror,-Wdeprecated-declarations]                                      
            applyPatternsAndFoldGreedily(getOperation(), std::move(patterns))))                                                          
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~                                                                                                 
            applyPatternsGreedily
```

Is there any way you could take a look at this?  Thanks!

MacDue commented:
As per the commit message, you should just replace call:

> For downstream users, there should be no behavioral change. Updating from the deprecated should just be a find and replace (e.g., find ./ -type f -exec sed -i 's|applyPatternsAndFoldGreedily|applyPatternsGreedily|g' {} \; variety) as the API hasn't changed between the two.

kazutakahirata commented:
> As per the commit message, you should just replace call:
> 
> > For downstream users, there should be no behavioral change. Updating from the deprecated should just be a find and replace (e.g., find ./ -type f -exec sed -i 's|applyPatternsAndFoldGreedily|applyPatternsGreedily|g' {} ; variety) as the API hasn't changed between the two.

Thanks!  I've checked in https://github.com/llvm/llvm-project/commit/9901906035c2be776674b25a3357b37a7dbbac68.
